Add AuthenticationFailure condition to dnszone controller

twiest opened the pull request:
Add AuthenticationFailure condition to dnszone controller
- [x] Add AuthenticationFailure condition to dnszone controller
- [x] Add unit tests
- [x] Manually test

staebler commented on pkg/controller/dnszone/dnszone_controller.go:
```diff
@@ -217,38 +218,76 @@ func (r *ReconcileDNSZone) reconcileDNSProvider(actuator Actuator, dnsZone *hive
-		dnsZone.Status.Conditions = conds
+	if accessDeniedCondsChanged {
+		dnsZone.Status.Conditions = accessDeniedConds
+		if err := r.Status().Update(context.Background(), dnsZone); err != nil {
```
Only call `r.Status().Update` once if either `accessDeniedCondsChanged` or `authFailureCondsChanged`.

twiest commented:
@staebler I believe I've addressed your feedback and I implemented the changes we discussed in standup.

I also finished the manual testing. I believe this PR is ready to go. PTAL.

staebler commented on pkg/controller/dnszone/awsactuator.go:
```diff
@@ -472,6 +472,53 @@ func (a *AWSActuator) Exists() (bool, error) {
+		return false
+	}
+
+	conds := []hivev1.DNSZoneCondition{}
```
```suggestion
	var conds []hivev1.DNSZoneCondition
```

staebler commented on pkg/controller/dnszone/awsactuator.go:
```diff
@@ -472,6 +472,53 @@ func (a *AWSActuator) Exists() (bool, error) {
+		)
+		failureMessage = "Credentials are lacking permissions"
+
+	case "InvalidSignatureException":
+		fallthrough
+	case "UnrecognizedClientException":
```
I think the following conveys the logic a bit better. The fallthough, at least to me, implies that the first case did something and should fall through to do more.
```suggestion
	case "InvalidSignatureException", "UnrecognizedClientException":
```

staebler commented on pkg/controller/dnszone/awsactuator.go:
```diff
@@ -472,6 +472,53 @@ func (a *AWSActuator) Exists() (bool, error) {
+		a.dnsZone.Status.Conditions = conds
+	}
+
+	a.logger.WithError(err).Warn(failureMessage)
```
This seems out of place here.

staebler commented on pkg/controller/dnszone/dnszone_controller.go:
```diff
@@ -217,38 +217,43 @@ func (r *ReconcileDNSZone) reconcileDNSProvider(actuator Actuator, dnsZone *hive
 	}
 
-	conds, changed := controllerutils.SetDNSZoneConditionWithChangeCheck(
+	accessDeniedConds, accessDeniedCondsChanged := controllerutils.SetDNSZoneConditionWithChangeCheck(
```
These conditions will only get set to false when there are no longer **any** errors. For example, let's say that there was initially an authentication error. The user fixes that but then there is an access denied error. The DNSZone will show that there is still an authentication error and that there is now **also** an access denied error.

twiest replied:
In your example, wouldn't fixing the authentication error cause authenticationFailureCondsChanged to become true?

I'm not seeing how the code works the way you describe.

twiest replied:
Ah, I get it. You're saying because we'd go into the ```if err != nil``` above and so this code wouldn't be reached. Good catch.

twiest commented:
@staebler I believe I've addressed your concerns. PTAL.

staebler commented on pkg/controller/dnszone/awsactuator.go:
```diff
@@ -472,6 +472,107 @@ func (a *AWSActuator) Exists() (bool, error) {
+		a.dnsZone.Status.Conditions,
+		hivev1.InsufficientCredentialsCondition,
+		corev1.ConditionFalse,
+		accessDeniedReason,
```
This is not the correct reason.

twiest replied:
Are you sure?

That's not my code, I just refactored it here.

staebler replied:
I am sure that it is the wrong reason. I do see that it is an existing problem and not something introduce in this PR.

twiest replied:
What should the reason be then?

@akhil-rane git blame says that you added this. Do you have input on what this should be?

staebler replied:
In this case, the condition is being set to False. The reason should not be "AccessDenied" because access was not denied.

staebler commented on pkg/controller/dnszone/awsactuator.go:
```diff
@@ -472,6 +472,107 @@ func (a *AWSActuator) Exists() (bool, error) {
+		return accessDeniedCondsChanged || authenticationFailureCondsChanged
+	}
+
+	awsErr, ok := err.(awserr.Error)
```
The InsufficientCredentials and AuthenticationFailure conditions should be set to false when the error is not an `awserr.Error`.

twiest replied:
Fixed.

staebler commented on pkg/controller/dnszone/awsactuator.go:
```diff
@@ -472,6 +472,110 @@ func (a *AWSActuator) Exists() (bool, error) {
+
+// SetConditionsForError sets conditions on the dnszone given a specific error. Returns true if conditions changed.
+func (a *AWSActuator) SetConditionsForError(err error) bool {
+	if err == nil {
```
This check is not necessary. It will be covered by the check for whether the error is a `awserr.Error`.

twiest commented:
@staebler @akhil-rane I addressed your comments, PTAL.

Specifically are the new `access*Reason` and `authentication*Reason` acceptable to you?
